[ARMv7] Support running testb3

justinmichaud opened the pull request:
#### 273f5243aa85af80bc22a9dbbd370efc61e726f6
<pre>
[ARMv7] Support running testb3
<a href="https://bugs.webkit.org/show_bug.cgi?id=276367">https://bugs.webkit.org/show_bug.cgi?id=276367</a>

Reviewed by OOPS.

This patch was written by Joseph Griego a while back. This does the following:

- Support running testb3 on armv7
- Allow dumping testb3 output, to ensure these changes don&apos;t affect other platforms

testb3 does not pass fully yet, but it does pass simple tests like testRotR.

I also added a list of skipped subtests, and confirmed on x86_64 and ARM64 (non-E)
that the generated B3 code is not changed.

To make these tests run, we mostly just abstract the calling convention away. Every
other change is mechanical.

Canonical link: <a href="https://commits.webkit.org/282881@main">https://commits.webkit.org/282881@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/154c5fedd93fd7e46131b7caeef7f3bdc454aba8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64533 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43899 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17130 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68556 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66652 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51642 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15420 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51920 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10449 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67601 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55844 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32541 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37270 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13223 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14014 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/57644 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59210 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13552 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70255 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/63777 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8480 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13059 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59245 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8514 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55933 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59421 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14236 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7013 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/688 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85538 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39711 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15087 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40789 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41972 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40532 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aoikonomopoulos commented on Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
```diff
@@ -1661,7 +1659,10 @@ class LowerToAir {
+                if (is32Bit() && value.value()->type() == Int64)
+                    arg = someArg(value.value());
+                else
+                    arg = tmp(value.value());
```
Minor nit: I think someArg can handle both cases here. Also is32Bit() should always be true in this file.

Constellation commented on Source/JavaScriptCore/b3/testb3.h:
```diff
@@ -117,7 +120,7 @@ extern Lock crashLock;
 #define CHECK_EQ(x, y) do { \
     auto __x = (x); \
-    auto __y = (y); \
+    decltype(__x) __y = (y); \
```
Why do we need this change?

justinmichaud replied:
This avoids having to add some casts for literals. As far as I can tell, it shouldn't implicitly convert anything, it just avoids having to add some extra casts to literals.

justinmichaud replied:
I fixed this

Constellation commented on Source/JavaScriptCore/b3/testb3_1.cpp:
```diff
@@ -166,21 +167,24 @@ void run(const TestConfig* config)
     addBitTests(config, tasks);
 
-    RUN(testShlArgs(1, 0));
+    RUN(testShlArgs(1, 0))
```
Why is it removing `;`?

justinmichaud replied:
No reason, I will revert the change.

Constellation commented on Source/JavaScriptCore/b3/testb3_failingArmTests.inc:
```diff
@@ -0,0 +1,258 @@
+"testInt64BImmToDoubleBitwiseCast",
```
What is this file?

justinmichaud replied:
This file is the list of skipped tests on armv7

Constellation commented on Source/JavaScriptCore/b3/testb3_7.cpp:
```diff
@@ -945,12 +978,13 @@ void testLICMControlDependentNotBackwardsDominant()
     generateLoopNotBackwardsDominant(
         proc, array,
         [&] (BasicBlock* loop, Value*) -> Value* {
+            auto arguments = cCallArgumentValues<unsigned *>(proc, loop);
```
Can you remove space between `unsigned` and `*`? And do that for all the other places?

justinmichaud replied:
Done!

ddegazio commented on Source/JavaScriptCore/b3/testb3_failingArmTests.inc:
Nit: maybe we can rename this file to `testb3_failingARMv7Tests.inc`? To avoid confusion with arm64.

ddegazio reviewed (approved):
r=me

jjgriego committed:
[ARMv7] Support running testb3
https://bugs.webkit.org/show_bug.cgi?id=276367

Reviewed by OOPS.

This patch was written by Joseph Griego a while back. This does the following:

- Support running testb3 on armv7
- Allow dumping testb3 output, to ensure these changes don't affect other platforms

testb3 does not pass fully yet, but it does pass simple tests like testRotR.

I also added a list of skipped subtests, and confirmed on x86_64 and ARM64 (non-E)
that the generated B3 code is not changed.

To make these tests run, we mostly just abstract the calling convention away. Every
other change is mechanical.

Canonical link: https://commits.webkit.org/282881@main
